File: app/src/main/java/tr/name/fatihdogan/books/service/SyncService.java
package tr.name.fatihdogan.books.service;

import android.app.IntentService;
import android.content.Intent;
import android.support.annotation.Nullable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import tr.name.fatihdogan.books.BaseApplication;
import tr.name.fatihdogan.books.api.GoogleBooksApi;
import tr.name.fatihdogan.books.api.response.Volume;
import tr.name.fatihdogan.books.api.response.Volumes;
import tr.name.fatihdogan.books.repository.Book;
import tr.name.fatihdogan.books.repository.BookDao;
import tr.name.fatihdogan.books.utils.StringUtils;

public class SyncService extends IntentService {

    private final GoogleBooksApi googleBooksApi;

    private final BookDao bookDao;

    public SyncService() {
        super("SyncService");
        googleBooksApi = BaseApplication.getAppComponent().googleBooksApi();
        bookDao = BaseApplication.getAppComponent().bookDao();
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        List<Volume> volumesList = new ArrayList<>();
        int total_count;
        int last_count;
        int last_index = 0;
        do {
            try {
                Volumes volumes = googleBooksApi.getMyVolumes(last_index).execute().body();
                if (volumes == null) {
                    return;
                    //TODO Handle error
                }
                total_count = volumes.totalItems;
                last_count = volumes.items == null ? 0 : volumes.items.size();
                last_index += last_count;
                if (volumes.items != null) {
                    volumesList.addAll(volumes.items);
                }
            } catch (IOException e) {
                //TODO Handle error
                e.printStackTrace();
                return;
            }

        } while (total_count > 0 && last_count > 0);

        if (volumesList.size() == total_count) {
            String[] ids = new String[volumesList.size()];
            for (int i = 0; i < volumesList.size(); i++)
                ids[i] = volumesList.get(i).id;
            bookDao.deleteOtherThanTheseIds(ids);
        }

        Book[] books = new Book[volumesList.size()];

        for (int i = 0; i < volumesList.size(); i++) {
            Volume volume = volumesList.get(i);
            Book book = bookDao.getById(volume.id);
            if (book == null) {
                book = new Book();
                book.setBookId(volume.id);
            }
            book.setOriginalAuthors(Arrays.asList(volume.volumeInfo.authors));
            book.setOriginalTitle(volume.volumeInfo.title);
            if (book.getAuthors() == null || book.getAuthors().size() == 0)
                book.setAuthors(book.getOriginalAuthors());

            if (StringUtils.isEmpty(book.getTitle()))
                book.setTitle(book.getOriginalTitle());

            if (StringUtils.isEmpty(book.getSortTitle()))
                book.setSortTitle(book.getTitle());

            books[i] = book;
        }

        bookDao.insertAll(books);
    }
}

File: app/src/main/java/com/fthdgn/books/service/SyncService.kt
package com.fthdgn.books.service

import android.app.IntentService
import android.content.Intent
import com.fthdgn.books.BaseApplication
import com.fthdgn.books.api.GoogleBooksApi
import com.fthdgn.books.api.response.Volume
import com.fthdgn.books.repository.Book
import com.fthdgn.books.repository.BookDao
import java.io.IOException
import java.util.*

class SyncService : IntentService("SyncService") {

    private val googleBooksApi: GoogleBooksApi = BaseApplication.getAppComponent().googleBooksApi()
    private val bookDao: BookDao = BaseApplication.getAppComponent().bookDao()

    override fun onHandleIntent(intent: Intent?) {
        val volumesList = ArrayList<Volume>()
        var totalCount: Int
        var lastCount: Int
        var lastIndex = 0
        do {
            try {
                val volumes = googleBooksApi.getMyVolumes(lastIndex).execute().body() ?: return
                //TODO Handle error
                totalCount = volumes.totalItems
                lastCount = if (volumes.items == null) 0 else volumes.items!!.size
                lastIndex += lastCount
                if (volumes.items != null) {
                    volumesList.addAll(volumes.items!!)
                }
            } catch (e: IOException) {
                //TODO Handle error
                e.printStackTrace()
                return
            }

        } while (totalCount > 0 && lastCount > 0)

        if (volumesList.size == totalCount) {
            val ids = volumesList.indices.map { volumesList[it].id }
            val ar = ids.toTypedArray()
            bookDao.deleteOtherThanTheseIds(*ar)
        }

        val books = arrayListOf<Book>()

        for (i in volumesList.indices) {
            val volume = volumesList[i]
            var book: Book? = bookDao.getById(volume.id)
            if (book == null) {
                book = Book(volume.id)
            }
            book.originalAuthors = Arrays.asList(*volume.volumeInfo!!.authors!!)
            book.originalTitle = volume.volumeInfo!!.title
            if (book.authors == null || book.authors!!.isEmpty())
                book.authors = book.originalAuthors

            if (book.title?.isEmpty() != false)
                book.title = book.originalTitle

            if (book.sortTitle?.isEmpty() != false)
                book.sortTitle = book.title

            books.add(book)
        }
        val ar = books.toTypedArray()

        bookDao.insertAll(*ar)
    }
}
